### PR TITLE
AK/LIbCore/LibCrypto/LibC:  Add AK::timing_safe_compare, and use it 

### PR DESCRIPTION
**Pull request by bgianfo**
In #12978 @jessicah brought up that we are not performing the `LibCore::Account` hash comparison with a constant time implementation. While it's not critical to use constant time for hash comparisons, it is considered best practice. So lets do that for `LibCore::Account` and then also fix up some cases in `LibCrypto` where we should be using a constant time comparison function as well.

A primer on constant time crypto that is useful: https://www.bearssl.org/constanttime.html

### CODE REVIEW

**Comment by bgianfo** [comment]
Actually, I'll make this a draft for now. 

I want to double check the disassembly across clang / gcc and make sure nothing is sneaking in.

File: AK/Memory.h [hunk 1]
```diff
@@ -50,6 +52,29 @@ inline void secure_zero(void* ptr, size_t size)
     asm volatile("" ::
                      : "memory");
 }
+
+// Naive implementation of a constant time buffer comparison function.
+// The goal being to not use any conditional branching so calls are
+// guarded against potential timing attacks.
+//
+// See OpenBSD's timingsafe_memcmp  for more advanced implementations.
```
Comment (alimpfard) [inline AK/Memory.h:60]: extra space after `timingsafe_memcmp`.

File: AK/Memory.h [hunk 2]
```diff
@@ -50,6 +52,29 @@ inline void secure_zero(void* ptr, size_t size)
...
+inline bool timing_safe_compare(const void* b1, const void* b2, size_t len)
+{
+    auto* c1 = reinterpret_cast<const char*>(b1);
+    auto* c2 = reinterpret_cast<const char*>(b2);
```
Comment (alimpfard) [inline AK/Memory.h:64]: `static_cast` should work fine here too?

File: Tests/AK/TestMemory.cpp [hunk 3]
```diff
@@ -66,3 +68,12 @@ TEST_CASE(kmp_two_chunks)
     EXPECT_EQ(result_2.value_or(9), 4u);
     EXPECT(!result_3.has_value());
 }
+
+TEST_CASE(timingsafe_compare)
```
Comment (alimpfard) [inline Tests/AK/TestMemory.cpp:72]: Idea for a future "make sure we don't branch on input data" I've had for a while that seems relevant here:
a memory "poison" API for UE (set all shadow bits to 0), plus an option to make UE quit on such branches; then `poison(a); posion(b); timing_safe_compare(a, b);` should cleanly execute (your current impl - as noted by the fixme - _can_ fail this test)

**Review by alimpfard — COMMENTED** [review]
impl looks good for a first impl, LibCrypto changes also LGTM.

File: Tests/AK/TestMemory.cpp (hunk 3 above)
Comment (bgianfo) [reply Tests/AK/TestMemory.cpp:72]: I like that idea, not sure I'll ever get around to it, but certainly sounds useful.

Comment (alimpfard) [reply Tests/AK/TestMemory.cpp:72]: lol same here, it'll happen eventually when we need it I guess

File: AK/Memory.h (hunk 1 above)
Comment (bgianfo) [reply AK/Memory.h]: Fixed

File: AK/Memory.h (hunk 2 above)
Comment (bgianfo) [reply AK/Memory.h]: Yup! Fixed

**Review by awesomekling — APPROVED** [review]

**Pull request merged**